app/next-client-app/lib/concept-utils.ts

```diff
@@ -0,0 +1,183 @@
+import { objToQuery } from "./client-utils";
+
+//function to map concept domain to an omop field
+export const mapConceptToOmopField = () => {
+  let omopTables: OmopTable[] | null = null;
+  // This const will return a function which will receive three params, and one is optional
+  // This optional one is the name of the omop table (check api/omoptables/ to confirm), so it should have the name omop_table_name: string, not table (misleading)
+  // This optional one also maybe not nessessary because we didn't pass it anyway in add_concept.tsx, so it will not be there by any means. And because of the reason on line 44, I suggest we can remove it, line 19 and lines 45-54
+  return async (
+    fields: OmopField[],
+    domain: string,
+    omop_table_name?: string
+  ) => {
+    // If omop_table_name is not provided (which will not be anyway), the function filters fields to get those that match the provided domain
+
+    if (!omop_table_name) {
+      // This condition is confusing and misleading, I suggest to remove it
+      // get omop fields that match specified domain
+      const mappedFields = fields.filter((field) => field.field == domain);
+      if (mappedFields.length < 2) {
+        return mappedFields[0]; // The code last time still worked, even though it had some errors, becasue it stopped here
+      }
+      //  If there are multiple matching fields, it retrieves OMOP tables data
+      if (!omopTables) {
+        omopTables = await getOmopTables();
+      }
+      // finds the field whose table is included in the m_allowed_tables list, which we will do again in the next step of add_conept.tsx, but we need it here.
+      let mappedTables: {
+        table: OmopTable | null;
+        field: OmopField;
+        isAllowed?: boolean; // Add this one here only for the purpose of filter the field whose table is included in the m_allowed_tables list or not
+      }[] = mappedFields.map((field) => ({
+        table: omopTables?.find((t) => t.id == field.table) ?? null,
+        field: field,
+      }));
+      mappedTables = mappedTables.map((val) => ({
+        ...val,
+        isAllowed: m_allowed_tables.includes(val.table?.table ?? ""),
+      }));
+      return mappedTables.find((val) => val.isAllowed == true)?.field; // Return the valid field
+    }
+    // If omop_table_name is provided, which will not happen (so we can remove the code below, I suppose), the function retrieves OMOP tables data ...
+    if (!omopTables) {
+      omopTables = await getOmopTables();
+    }
+    // ... then finds the field that matches both the provided omop_table_name and passed domain
+    let mappedTable =
+      omopTables.find((t) => t.table == omop_table_name) ?? null;
+    const mappedField = fields.find(
+      (f) => f.table == mappedTable?.id && f.field == domain
+    );
+    return mappedField; // return the matching field
+  };
+};
+
+// function to save mapping rules copying python implementation
```

Comment:
For example on simplifying, I think we would definitely like to avoid anything that is copying the existing python implementation.